stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -46,81 +53,102 @@ namespace stan {
+       * Reverse mode differentiation
        * algorithm refernce:
        *
-       * Mike Giles. An extended collection of matrix
-       * derivative results for forward and reverse mode AD.
-       * Jan. 2008.
+       * Iain Murray: Differentiation of
+       * the Cholesky decomposition, 2016.
        *
-       * Note algorithm  as laid out in Giles is
-       * row-major, so Eigen::Matrices are explicitly storage
-       * order RowMajor, whereas Eigen defaults to
-       * ColumnMajor. Also note algorithm
-       * starts by calculating the adjoint for
-       * A(M_ - 1, M_ - 1), hence pos on line 94 is decremented
-       * to start at pos = M_ * (M_ + 1) / 2.
-       * */
+       */
       virtual void chain() {
-        using Eigen::Matrix;
-        using Eigen::RowMajor;
-        Matrix<double, -1, -1, RowMajor> adjL(M_, M_);
-        Matrix<double, -1, -1, RowMajor> LA(M_, M_);
-        Matrix<double, -1, -1, RowMajor> adjA(M_, M_);
+        using Eigen::MatrixXd;
+        using Eigen::Lower;
+        using Eigen::Block;
+        using Eigen::Upper;
+        using Eigen::StrictlyUpper;
+        MatrixXd Lbar(M_, M_);
+        MatrixXd L(M_, M_);
+
+        Lbar.setZero();
+        L.setZero();
         size_t pos = 0;
-        for (size_type i = 0; i < M_; ++i) {
-          for (size_type j = 0; j <= i; ++j) {
-            adjL.coeffRef(i, j) = variRefL_[pos]->adj_;
-            LA.coeffRef(i, j) = variRefL_[pos]->val_;
+        for (size_type j = 0; j < M_; ++j) {
+          for (size_type i = j; i < M_; ++i) {
+            Lbar.coeffRef(i, j) = variRefL_[pos]->adj_;
+            L.coeffRef(i, j) = variRefL_[pos]->val_;
             ++pos;
           }
         }
 
-        --pos;
-        for (int i = M_ - 1; i >= 0; --i) {
-          for (int j = i; j >= 0; --j) {
-            if (i == j) {
-              adjA.coeffRef(i, j) = 0.5 * adjL.coeff(i, j)
-                / LA.coeff(i, j);
-            } else {
-              adjA.coeffRef(i, j) = adjL.coeff(i, j)
-                / LA.coeff(j, j);
-              adjL.coeffRef(j, j) -= adjL.coeff(i, j)
-                * LA.coeff(i, j) / LA.coeff(j, j);
-            }
-            for (int k = j - 1; k >=0; --k) {
-              adjL.coeffRef(i, k) -= adjA.coeff(i, j)
-                * LA.coeff(j, k);
-              adjL.coeffRef(j, k) -= adjA.coeff(i, j)
-                * LA.coeff(i, k);
-            }
-            variRefA_[pos--]->adj_ += adjA.coeffRef(i, j);
+        for (int k = M_; k > 0; k -= block_size_) {
+          int j = std::max(0, k - block_size_);
+          Block_ R = L.block(j, 0, k - j, j);
+          Block_ D = L.block(j, j, k - j, k - j);
+          Block_ B = L.block(k, 0, M_ - k, j);
+          Block_ C = L.block(k, j, M_ - k, k - j);
+          Block_ Rbar = Lbar.block(j, 0, k - j, j);
+          Block_ Dbar = Lbar.block(j, j, k - j, k - j);
+          Block_ Bbar = Lbar.block(k, 0, M_ - k, j);
+          Block_ Cbar = Lbar.block(k, j, M_ - k, k - j);
```

Comment:
Yeah, the tests made updating this function much easier than it otherwise would have been.